woocommerce/woocommerce-ios
# Shipping Labels: Update contact button in the warning banner of Ship To form with options to email or message customers

Fixes: #4791

## Description

When a destination address for shipping labels cannot be verified, we're showing a banner with the option to contact customer. This PR updates the "Contact Customer" action so the merchant can email, call, or message the customer (if email and phone are available) instead of just calling them.

## Changes

* The customer's email address comes from their billing address (there is no shipping email), so the billing email is passed from the Shipping Label Form to the Shipping Label Address Form, where it's used for the banner when editing the Ship To (destination) address.
* The banner now has an action sheet with options to Email, Call, or Message the customer.
   * As before, if there is no phone or email the option to Contact Customer will not appear in the banner.
   * If there is only a phone number, the action sheet will only show the Call or Message options.
   * If there is only an email address, the action sheet will only show the Email option.
* The banner factory only needs to know if there is a phone/email, so it was updated to accept `Bool` parameters `hasPhone` and `hasEmail` and uses those to determine what banner text and actions to display.
* Added unit tests for the `ShippingLabelAddressTopBannerFactory`, to check that the correct number of actions appear on the banner.

Phone & Email|Phone only|Email only
-|-|-
![IMG_7850](https://user-images.githubusercontent.com/8658164/137486143-e362bd16-68fe-452c-a43f-b13b2eb8e8a8.PNG)|![IMG_7852](https://user-images.githubusercontent.com/8658164/137486126-1b65b7f3-99a2-4158-8f3f-af4dc535234e.PNG)|![IMG_7851](https://user-images.githubusercontent.com/8658164/137486137-9c7e3c0f-b238-4eb5-8812-bd4484bb2018.PNG)

## Testing

1. Make sure your test store has the WCShip plugin installed and activated.
2. Create an order that is eligible for shipping labels, with an invalid shipping address, a billing email address, and a billing or shipping phone number.
3. Build the app on a real device with the Mail app (so you can test the contact options).
4. Select the order and tap Create Shipping Label.
5. Confirm the Ship From address.
6. Select the Ship To section, and confirm it opens the address edit form with the warning banner at the top of the screen.
7. Tap the "Contact Customer" button and confirm an action sheet opens with the expected contact options.
8. Select each contact option and confirm each one works as expected (giving you the option to call the phone number, send a text message to the phone number, or email the email address).
9. Remove the phone number, tap Done, and confirm that the Contact Customer action sheet now only shows the Email option.
10. Remove the email address from the order (this must be done in wp-admin) or select an order with no billing email. Go back to the order detail, pull to refresh the order details, and repeat steps 4-8 above, confirming that no Email option appears.

## Submitter Checklist

Update release notes:
